analyzers/src/SonarAnalyzer.Common/Helpers/AspNetMvcHelper.cs

```diff
@@ -44,18 +44,22 @@ public static class AspNetMvcHelper
         /// </summary>
         public static bool IsControllerMethod(this IMethodSymbol methodSymbol) =>
             methodSymbol.MethodKind == MethodKind.Ordinary
-            && methodSymbol.GetEffectiveAccessibility() == Accessibility.Public
+            && methodSymbol.DeclaredAccessibility == Accessibility.Public
```

Comment:
@martin-strecker-sonarsource I changed this condition to cover the UT cases such as
```csharp
    protected class ProtectedController : Controller            // Noncompliant
    {
        [Route("/Index1")] // Secondary
        public IActionResult Index() => View();
    }
```

But I'm not sure anymore. Unless it's a private class I think for internal and protected we should still consider these actions right?

Reply:
I would say that controller that are nested classes are not discovered by the MVC routing mechanism, but I haven't tried that. According to the docs

> A controller is an instantiable class, usually [public](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/keywords/public), in which at least one of the following conditions is true:

* The class name is suffixed with Controller.
* The class inherits from a class whose name is suffixed with Controller.
* The [Controller] attribute is applied to the class.

> A controller class must not have an associated [NonController] attribute.

[Source](https://learn.microsoft.com/en-us/aspnet/core/mvc/controllers/actions?view=aspnetcore-8.0#what-is-a-controller)

So I would say, public (and maybe internal) should be enough. If you can make nested controllers work, you can also leave a comment in the tests and here, how this is working.

Reply:
Just tested - only actions in public controllers are discoverable.
I'll update the UTs accordingly.